unison-cli/src/Unison/Codebase/Editor/HandleInput/Update.hs

```diff
@@ -189,12 +201,324 @@ handleUpdate input optionalPatch requestedNames = do
+
+    let termRefToNames :: TermReferenceId -> Set Symbol
+        termRefToNames =
+          Set.map Name.toVar . Names.namesForReferent slurpCheckNames . Referent.fromTermReferenceId
+
+    let nameToTermRefs :: Symbol -> Set TermReference
+        nameToTermRefs = Names.refTermsNamed slurpCheckNames . Name.unsafeFromVar
+
+    pure (slurp, termRefToNames, nameToTermRefs)
+
+  slurp1 <- do
+    Cli.Env {codebase} <- ask
+    unisonFile0 <- Cli.expectLatestTypecheckedFile
+
+    -- Here, we identify whether there are any "implicit terms", which are terms that should be brought into the latest
+    -- typechecked Unsion file and re-typechecked, which also re-discovers components.
+    --
+    -- The running example here will be one in which the current namespace already has the following terms stored.
+    -- (For simplicity, we will ignore the limitation that mutually recursive definitions must be top-level arrows;
+    -- pretend this is Haskell's let).
+    --
+    --   ping = pong + 1   (reference = #pingpong.ping)
+    --   pong = ping + 2   (reference = #pingpong.pong)
+    --   wham = pong + 3   (reference = #wham)
+    --
+    -- The user re-defines "ping" in their scratch file thus:
+    --
+    --   ping = wham + 4   (reference = #newping)
+    --
+    -- Note that, pre-update, we had two components [ping,pong] and [wham]. But after the update, since the new `ping`
+    -- refers to the old `wham`, which refers to the old `pong`, which refers to the old `ping`, we really want to end
+    -- up with a single component in the end, [ping,pong,wham].
+
+    -- First, compute an initial slurp, which will identify the initial set of definitions we are updating ({"ping"}).
+    -- Also, this will be the slurp that we fall back on, in case re-typechecking another Unison file with implicit
+    -- terms in it fails.
+    let slurp0 = slurp unisonFile0
+
+    -- Get the set of names that are being updated.
+    --
+    -- Running example:
+    --
+    --   { "ping" }
+    let namesBeingUpdated :: Set Symbol
+        namesBeingUpdated =
+          SC.terms (Slurp.updates slurp0)
+
+    -- Associate each such name with the set of "old" references (already in the codebase) that it's associated with.
+    --
+    -- Running example:
+    --
+    --   "ping" => { #pingpong.ping }
+    let updatedNameToOldRefs :: Map Symbol (Set TermReference)
+        updatedNameToOldRefs =
+          Map.fromSet nameToTermRefs namesBeingUpdated
+
+    -- Identify all of the implicit terms, which are:
+    --
+    --   - Both:
+    --     - Either:
+    --         - (A) A component-mate of a term that's being updated.
+    --         - Both:
+    --           - (B) A dependent of a term that's being updated.
+    --           - (C) A dependency of the new term.
+    --     - (D) Not being updated.
+    --
+    -- FIXME Additionally, we have two more temporary requirements.
```

Comment:
`FIXME`

Reply:
Yeah, it's being tracked in the minimegaticket